File: src/daemon/kuper-auth.ts
import { mkdir, readFile, writeFile } from 'node:fs/promises';
import path from 'node:path';
import type { Browser, Page } from 'playwright';
import { fileExists } from './litres-auth.js';
import { launchCamoufoxBrowser } from '../playwright/browser-session.js';

export const DEFAULT_KUPER_STORAGE_STATE = '/root/.openclaw/workspace/tmp/sberid-login/kuper/storage-state.json';
export const DEFAULT_KUPER_BOOTSTRAP_OUT_DIR = '/root/.openclaw/workspace/tmp/sberid-login/kuper';
export const DEFAULT_KUPER_BOOTSTRAP_ENTRY_URL = 'https://kuper.ru/';
export const DEFAULT_SBER_COOKIES_PATH = '/root/.openclaw/workspace/sber-cookies.json';
export const REPO_OWNED_KUPER_BOOTSTRAP = 'repo:src/daemon/kuper-auth.ts';

export interface KuperBootstrapAttemptResult {
  attempted: boolean;
  ok: boolean;
  status:
    | 'not_attempted'
    | 'reused_existing_state'
    | 'not_applicable'
    | 'skipped_missing_cookies'
    | 'redirected_to_sberid'
    | 'handoff_required'
    | 'state_refreshed'
    | 'completed_without_auth'
    | 'failed';
  handoffRequired: boolean;
  redirectedToSberId: boolean;
  bootstrapFailed: boolean;
  scriptPath: string | null;
  statePath: string | null;
  outDir: string | null;
  finalUrl: string | null;
  rawStatus: string | null;
  errorMessage: string | null;
}

async function ensureDir(dir: string): Promise<void> {
  await mkdir(dir, { recursive: true });
}

async function saveBodyText(page: Page, file: string): Promise<string> {
  const text = await page.locator('body').innerText().catch(() => '');
  await writeFile(file, text || '', 'utf8');
  return text || '';
}

async function maybeScreenshot(page: Page, file: string, enabled: boolean, screenshots: string[]): Promise<void> {
  if (!enabled) return;
  await page.screenshot({ path: file, fullPage: true });
  screenshots.push(file);
}

export async function runIntegratedKuperBootstrap(input: {
  storageStatePath: string | null;
  cookiesPath?: string;
  outDir?: string;
  headed?: boolean;
  debugScreenshots?: boolean;
}): Promise<KuperBootstrapAttemptResult> {
  const cookiesPath = path.resolve(input.cookiesPath ?? DEFAULT_SBER_COOKIES_PATH);
  const outDir = path.resolve(input.outDir ?? DEFAULT_KUPER_BOOTSTRAP_OUT_DIR);

  if (!(await fileExists(cookiesPath))) {
    return {
      attempted: true,
      ok: false,
      status: 'skipped_missing_cookies',
      handoffRequired: false,
      redirectedToSberId: false,
      bootstrapFailed: true,
      scriptPath: REPO_OWNED_KUPER_BOOTSTRAP,
      statePath: input.storageStatePath,
      outDir,
      finalUrl: null,
      rawStatus: null,
      errorMessage: 'Sber cookies file is missing'
    };
  }

  const statePath = path.resolve(input.storageStatePath ?? DEFAULT_KUPER_STORAGE_STATE);
  const screenshots: string[] = [];
  await ensureDir(outDir);
  await ensureDir(path.dirname(statePath));

  const cookies = JSON.parse(await readFile(cookiesPath, 'utf8')) as Array<{
    name: string;
    value: string;
    domain: string;
    path: string;
    expires?: number;
    httpOnly?: boolean;
    secure?: boolean;
    sameSite?: 'Strict' | 'Lax' | 'None';
  }>;

  let stopCamoufox: (() => void) | null = null;
  let browser: Browser | null = null;
  let page: Page | null = null;

  try {
    const launched = await launchCamoufoxBrowser();
    browser = launched.browser;
    stopCamoufox = launched.stop;

    const reusedSavedState = await fileExists(statePath);
    const context = reusedSavedState
      ? await browser.newContext({ viewport: { width: 1440, height: 1200 }, storageState: statePath })
      : await browser.newContext({ viewport: { width: 1440, height: 1200 } });
    page = await context.newPage();

    await context.addCookies(cookies as Parameters<typeof context.addCookies>[0]);
    await context.storageState({ path: statePath });

    await page.goto(DEFAULT_KUPER_BOOTSTRAP_ENTRY_URL, { waitUntil: 'domcontentloaded', timeout: 120000 });

    // Wait up to 30s for anti-bot JS challenge to resolve and redirect back to kuper.ru
    await page.waitForURL(
      (url) => !/hcheck=/i.test(url.toString()) && !/\/xpvnsulc\//i.test(url.toString()),
      { timeout: 30000 }
    ).catch(() => {});

    await page.waitForTimeout(2000);
    await maybeScreenshot(page, path.join(outDir, '01-home.png'), Boolean(input.debugScreenshots), screenshots);

    // Check for anti-bot challenge
    const afterGotoUrl = page.url();
    if (/hcheck=/i.test(afterGotoUrl) || /\/xpvnsulc\//i.test(afterGotoUrl)) {
      await saveBodyText(page, path.join(outDir, 'page.txt'));
      await context.storageState({ path: statePath });
      return {
        attempted: true,
        ok: false,
        status: 'handoff_required',
        handoffRequired: true,
        redirectedToSberId: false,
        bootstrapFailed: false,
        scriptPath: REPO_OWNED_KUPER_BOOTSTRAP,
        statePath,
        outDir,
        finalUrl: afterGotoUrl,
        rawStatus: 'anti_bot_challenge',
        errorMessage: 'Anti-bot challenge detected on kuper.ru even with camoufox'
      };
    }

    // Check if already authenticated (kuper.ru session cookies might be valid)
    const homeText = await page.locator('body').innerText().catch(() => '');
    const homeLowered = homeText.toLowerCase();
    const alreadyAuth = /профил|выйти|личный кабинет/i.test(homeLowered);
    if (alreadyAuth) {
      await context.storageState({ path: statePath });
      return {
        attempted: true,
        ok: true,
        status: 'state_refreshed',
        handoffRequired: false,
        redirectedToSberId: false,
        bootstrapFailed: false,
        scriptPath: REPO_OWNED_KUPER_BOOTSTRAP,
        statePath,
        outDir,
        finalUrl: afterGotoUrl,
        rawStatus: 'already-authenticated',
        errorMessage: null
      };
    }

    // Find and click the login button — race against anti-bot redirect
    const loginBtn = page.locator('text=Войти').first();
    try {
      await loginBtn.waitFor({ state: 'visible', timeout: 25000 });
    } catch {
      // Check if anti-bot redirect happened during the wait
      const urlAfterWait = page.url();
      if (/hcheck=/i.test(urlAfterWait) || /\/xpvnsulc\//i.test(urlAfterWait)) {
        await saveBodyText(page, path.join(outDir, 'page.txt'));
        await context.storageState({ path: statePath });
        return {
          attempted: true,
          ok: false,
          status: 'handoff_required',
          handoffRequired: true,
          redirectedToSberId: false,
          bootstrapFailed: false,
          scriptPath: REPO_OWNED_KUPER_BOOTSTRAP,
          statePath,
          outDir,
          finalUrl: urlAfterWait,
          rawStatus: 'anti_bot_challenge',
          errorMessage: 'Anti-bot JS challenge blocked kuper.ru access — IP likely flagged as datacenter'
        };
      }
      throw new Error(`Login button not found: ${(loginBtn as unknown as Error).toString()}`);
    }

    const beforeClickUrl = page.url();
    await Promise.allSettled([
      page.waitForURL(/id\.sber\.ru|kuper\.ru/i, { timeout: 20000 }),
      loginBtn.click({ timeout: 10000 })
    ]);

    await page.waitForLoadState('domcontentloaded', { timeout: 30000 }).catch(() => {});
    await page.waitForTimeout(3000);

    const finalUrl = page.url();
    const text = await saveBodyText(page, path.join(outDir, 'page.txt'));
    await maybeScreenshot(page, path.join(outDir, '02-after-login-click.png'), Boolean(input.debugScreenshots), screenshots);
    await context.storageState({ path: statePath });

    const lowered = text.toLowerCase();
    const redirectedToSberId = /id\.sber\.ru/i.test(finalUrl);
    const maybeAuthenticated =
      /профил|выйти|личный кабинет|мой профиль/i.test(lowered) ||
      (!redirectedToSberId && finalUrl !== beforeClickUrl);
    const stateExists = await fileExists(statePath);

    if (redirectedToSberId) {
      return {
        attempted: true,
        ok: true,
        status: 'redirected_to_sberid',
        handoffRequired: true,
        redirectedToSberId: true,
        bootstrapFailed: false,
        scriptPath: REPO_OWNED_KUPER_BOOTSTRAP,
        statePath,
        outDir,
        finalUrl,
        rawStatus: 'redirected-to-sberid',
        errorMessage: null
      };
    }

    if (maybeAuthenticated) {
      return {
        attempted: true,
        ok: true,
        status: stateExists ? 'state_refreshed' : 'completed_without_auth',
        handoffRequired: false,
        redirectedToSberId: false,
        bootstrapFailed: false,
        scriptPath: REPO_OWNED_KUPER_BOOTSTRAP,
        statePath,
        outDir,
        finalUrl,
        rawStatus: 'maybe-authenticated',
        errorMessage: null
      };
    }

    return {
      attempted: true,
      ok: stateExists,
      status: stateExists ? 'state_refreshed' : 'completed_without_auth',
      handoffRequired: false,
      redirectedToSberId: false,
      bootstrapFailed: !stateExists,
      scriptPath: REPO_OWNED_KUPER_BOOTSTRAP,
      statePath,
      outDir,
      finalUrl,
      rawStatus: finalUrl === beforeClickUrl ? 'loaded' : 'navigated',
      errorMessage: stateExists ? null : 'Bootstrap finished without producing a reusable state file'
    };
  } catch (error) {
    if (page) {
      const errorShot = path.join(outDir, 'error.png');
      await page.screenshot({ path: errorShot, fullPage: true }).catch(() => {});
      if (!screenshots.includes(errorShot)) screenshots.push(errorShot);
    }

    return {
      attempted: true,
      ok: false,
      status: 'failed',
      handoffRequired: false,
      redirectedToSberId: false,
      bootstrapFailed: true,
      scriptPath: REPO_OWNED_KUPER_BOOTSTRAP,
      statePath: input.storageStatePath,
      outDir,
      finalUrl: page?.url() ?? null,
      rawStatus: null,
      errorMessage: error instanceof Error ? error.message : String(error)
    };
  } finally {
    await browser?.close().catch(() => {});
    stopCamoufox?.();
  }
}

File: browser_platform/daemon/kuper_auth.py
from __future__ import annotations
import asyncio
import json
import re
from pathlib import Path

DEFAULT_KUPER_STORAGE_STATE = '/root/.openclaw/workspace/tmp/sberid-login/kuper/storage-state.json'
DEFAULT_KUPER_BOOTSTRAP_OUT_DIR = '/root/.openclaw/workspace/tmp/sberid-login/kuper'
DEFAULT_KUPER_BOOTSTRAP_ENTRY_URL = 'https://kuper.ru/'
DEFAULT_SBER_COOKIES_PATH = '/root/.openclaw/workspace/sber-cookies.json'
REPO_OWNED_KUPER_BOOTSTRAP = 'repo:browser_platform/daemon/kuper_auth.py'


def _file_exists(path: str) -> bool:
    return Path(path).exists()


def _ensure_dir(directory: str) -> None:
    Path(directory).mkdir(parents=True, exist_ok=True)


async def _save_body_text(page, file_path: str) -> str:
    try:
        text = await page.locator('body').inner_text()
    except Exception:
        text = ''
    Path(file_path).write_text(text or '', 'utf-8')
    return text or ''


async def run_integrated_kuper_bootstrap(
    *,
    storage_state_path: str | None = None,
    cookies_path: str | None = None,
    out_dir: str | None = None,
    headed: bool = False,
    debug_screenshots: bool = False,
) -> dict:
    from ..playwright.browser_session import launch_camoufox_browser

    cookies_file = str(Path(cookies_path or DEFAULT_SBER_COOKIES_PATH).resolve())
    resolved_out_dir = str(Path(out_dir or DEFAULT_KUPER_BOOTSTRAP_OUT_DIR).resolve())

    if not _file_exists(cookies_file):
        return {
            'attempted': True, 'ok': False, 'status': 'skipped_missing_cookies',
            'handoffRequired': False, 'redirectedToSberId': False, 'bootstrapFailed': True,
            'scriptPath': REPO_OWNED_KUPER_BOOTSTRAP, 'statePath': storage_state_path,
            'outDir': resolved_out_dir, 'finalUrl': None, 'rawStatus': None,
            'errorMessage': 'Sber cookies file is missing',
        }

    state_path = str(Path(storage_state_path or DEFAULT_KUPER_STORAGE_STATE).resolve())
    screenshots: list = []
    _ensure_dir(resolved_out_dir)
    _ensure_dir(str(Path(state_path).parent))

    cookies = json.loads(Path(cookies_file).read_text('utf-8'))

    stop_camoufox = None
    browser = None
    page = None

    try:
        launched = await launch_camoufox_browser()
        browser = launched['browser']
        stop_camoufox = launched['stop']

        reused_saved_state = _file_exists(state_path)
        ctx_opts: dict = {'viewport': {'width': 1440, 'height': 1200}}
        if reused_saved_state:
            ctx_opts['storage_state'] = state_path
        context = await browser.new_context(**ctx_opts)
        page = await context.new_page()

        await context.add_cookies(cookies)
        await context.storage_state(path=state_path)

        await page.goto(DEFAULT_KUPER_BOOTSTRAP_ENTRY_URL, wait_until='domcontentloaded', timeout=120000)

        try:
            await page.wait_for_url(
                lambda url: not re.search(r'hcheck=|/xpvnsulc/', url, re.IGNORECASE),
                timeout=30000,
            )
        except Exception:
            pass

        await page.wait_for_timeout(2000)
        if debug_screenshots:
            shot = str(Path(resolved_out_dir) / '01-home.png')
            await page.screenshot(path=shot, full_page=True)
            screenshots.append(shot)

        after_goto_url = page.url
        if re.search(r'hcheck=|/xpvnsulc/', after_goto_url, re.IGNORECASE):
            await _save_body_text(page, str(Path(resolved_out_dir) / 'page.txt'))
            await context.storage_state(path=state_path)
            return {
                'attempted': True, 'ok': False, 'status': 'handoff_required',
                'handoffRequired': True, 'redirectedToSberId': False, 'bootstrapFailed': False,
                'scriptPath': REPO_OWNED_KUPER_BOOTSTRAP, 'statePath': state_path,
                'outDir': resolved_out_dir, 'finalUrl': after_goto_url,
                'rawStatus': 'anti_bot_challenge',
                'errorMessage': 'Anti-bot challenge detected on kuper.ru even with camoufox',
            }

        try:
            home_text = await page.locator('body').inner_text()
        except Exception:
            home_text = ''
        home_lowered = home_text.lower()
        already_auth = bool(re.search(r'профил|выйти|личный кабинет', home_lowered, re.IGNORECASE))

        if already_auth:
            await context.storage_state(path=state_path)
            return {
                'attempted': True, 'ok': True, 'status': 'state_refreshed',
                'handoffRequired': False, 'redirectedToSberId': False, 'bootstrapFailed': False,
                'scriptPath': REPO_OWNED_KUPER_BOOTSTRAP, 'statePath': state_path,
                'outDir': resolved_out_dir, 'finalUrl': after_goto_url,
                'rawStatus': 'already-authenticated', 'errorMessage': None,
            }

        login_btn = page.locator('text=Войти').first
        try:
            await login_btn.wait_for(state='visible', timeout=25000)
        except Exception:
            url_after_wait = page.url
            if re.search(r'hcheck=|/xpvnsulc/', url_after_wait, re.IGNORECASE):
                await _save_body_text(page, str(Path(resolved_out_dir) / 'page.txt'))
                await context.storage_state(path=state_path)
                return {
                    'attempted': True, 'ok': False, 'status': 'handoff_required',
                    'handoffRequired': True, 'redirectedToSberId': False, 'bootstrapFailed': False,
                    'scriptPath': REPO_OWNED_KUPER_BOOTSTRAP, 'statePath': state_path,
                    'outDir': resolved_out_dir, 'finalUrl': url_after_wait,
                    'rawStatus': 'anti_bot_challenge',
                    'errorMessage': 'Anti-bot JS challenge blocked kuper.ru access — IP likely flagged as datacenter',
                }
            raise Exception('Login button not found')

        before_click_url = page.url
        await asyncio.gather(
            page.wait_for_url(
                lambda u: bool(re.search(r'id\.sber\.ru|kuper\.ru', u, re.IGNORECASE)),
                timeout=20000,
            ),
            login_btn.click(timeout=10000),
            return_exceptions=True,
        )

        try:
            await page.wait_for_load_state('domcontentloaded', timeout=30000)
        except Exception:
            pass
        await page.wait_for_timeout(3000)

        final_url = page.url
        text = await _save_body_text(page, str(Path(resolved_out_dir) / 'page.txt'))
        if debug_screenshots:
            shot = str(Path(resolved_out_dir) / '02-after-login-click.png')
            await page.screenshot(path=shot, full_page=True)
            screenshots.append(shot)
        await context.storage_state(path=state_path)

        lowered = text.lower()
        redirected_to_sberid = bool(re.search(r'id\.sber\.ru', final_url, re.IGNORECASE))
        maybe_authenticated = (
            bool(re.search(r'профил|выйти|личный кабинет|мой профиль', lowered, re.IGNORECASE)) or
            (not redirected_to_sberid and final_url != before_click_url)
        )
        state_exists = _file_exists(state_path)

        if redirected_to_sberid:
            return {
                'attempted': True, 'ok': True, 'status': 'redirected_to_sberid',
                'handoffRequired': True, 'redirectedToSberId': True, 'bootstrapFailed': False,
                'scriptPath': REPO_OWNED_KUPER_BOOTSTRAP, 'statePath': state_path,
                'outDir': resolved_out_dir, 'finalUrl': final_url,
                'rawStatus': 'redirected-to-sberid', 'errorMessage': None,
            }

        if maybe_authenticated:
            return {
                'attempted': True, 'ok': True,
                'status': 'state_refreshed' if state_exists else 'completed_without_auth',
                'handoffRequired': False, 'redirectedToSberId': False, 'bootstrapFailed': False,
                'scriptPath': REPO_OWNED_KUPER_BOOTSTRAP, 'statePath': state_path,
                'outDir': resolved_out_dir, 'finalUrl': final_url,
                'rawStatus': 'maybe-authenticated', 'errorMessage': None,
            }

        return {
            'attempted': True, 'ok': state_exists,
            'status': 'state_refreshed' if state_exists else 'completed_without_auth',
            'handoffRequired': False, 'redirectedToSberId': False, 'bootstrapFailed': not state_exists,
            'scriptPath': REPO_OWNED_KUPER_BOOTSTRAP, 'statePath': state_path,
            'outDir': resolved_out_dir, 'finalUrl': final_url,
            'rawStatus': 'loaded' if final_url == before_click_url else 'navigated',
            'errorMessage': None if state_exists else 'Bootstrap finished without producing a reusable state file',
        }

    except Exception as exc:
        if page:
            error_shot = str(Path(resolved_out_dir) / 'error.png')
            try:
                await page.screenshot(path=error_shot, full_page=True)
                screenshots.append(error_shot)
            except Exception:
                pass
        return {
            'attempted': True, 'ok': False, 'status': 'failed',
            'handoffRequired': False, 'redirectedToSberId': False, 'bootstrapFailed': True,
            'scriptPath': REPO_OWNED_KUPER_BOOTSTRAP, 'statePath': storage_state_path,
            'outDir': resolved_out_dir, 'finalUrl': page.url if page else None,
            'rawStatus': None, 'errorMessage': str(exc),
        }
    finally:
        tasks = []
        if browser:
            tasks.append(asyncio.ensure_future(_safe_close(browser)))
        if stop_camoufox:
            tasks.append(asyncio.ensure_future(_safe_stop(stop_camoufox)))
        if tasks:
            await asyncio.gather(*tasks, return_exceptions=True)


async def _safe_close(obj) -> None:
    try:
        await obj.close()
    except Exception:
        pass


async def _safe_stop(fn) -> None:
    try:
        result = fn()
        if asyncio.iscoroutine(result):
            await result
    except Exception:
        pass
